File: sorting/countingsort/c/countingsort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../../../utils/utils.h"

void counting_sort(int a[], int n) {
    int *b = (int*)malloc(n*sizeof(int)); // buffer array
    memset(b, 0, sizeof(int)*n);
    int max = a[0];
    for (int i = 1; i < n; ++i)
        if (max < a[i])
            max = a[i];
    ++max;
    int *c = (int*)malloc(sizeof(int)*max); // count array
    memset(c, 0, sizeof(int)*max);
    for (int i = 0; i < n; ++i)
        ++c[a[i]];

    for (int i = 1; i < max; ++i)
        c[i] += c[i-1];

    for (int i = n-1; i >= 0; --i) {
        b[c[a[i]]-1] = a[i];
        --c[a[i]];
    }
    for (int i = 0; i < n; ++i)
        a[i] = b[i];
    free(b);
    free(c);
}

int main(void) {
    int a[20];
    random_array(a, 20);
    print_array(a, 20);
    counting_sort(a, 20);
    print_array(a, 20);

    return 0;
}

File: sorting/countingsort/java/CountingSort.java

public class CountingSort {
    public static int[] countingSort(int[] A) {
        int[] B = new int[A.length];
        int max = A[0];
        for (int i = 1; i < A.length; ++i)
            if (max < A[i])
                max = A[i];
        countingSort(A, B, ++max);
        return B;
    }

    private static void countingSort(int[] A, int[] B, int k) {
        int[] C = new int[k];
        for (int i = 0; i < A.length; ++i)
            ++C[A[i]];
        for (int i = 1; i < k; ++i)
            C[i] += C[i-1];
        for (int i = A.length - 1; i >= 0; --i) {
            B[C[A[i]]-1] = A[i];
            --C[A[i]];
        }
    }
    public static void main(String[] argv) {
        int[] a = CountingSort.countingSort(new int[]{16, 4, 10, 12, 32, 55, 0,
                                                      99, 13, 88, 44});
        for (int i = 0; i < a.length; ++i)
            System.out.print(" "+a[i]);
        System.out.println();
    }
}
